php/libraries/DefaultFilesUploadHandler.class.inc

```diff
@@ -0,0 +1,367 @@
+     * Accessor for field $_moveDirPath.
+     *
+     * @return string value for field $_moveDirPath.
+     */
+    public function getMoveDirPath() : string
+    {
+        return $this->_moveDirPath;
+    }
+
+    /**
+     * Accessor for field $_dirGroup.
+     *
+     * @return int value for field $_dirGroup.
+     */
+    public function getDirGroup() : string
+    {
+        return $this->_dirGroup;
+    }
+
+    /**
+     * Accessor for field $_movedFiles.
+     *
+     * @return array the final location of all files uploaded on the
+     *               server. Each key is the name of a LorisFirm field
+     *               used to upload a file and its corresponding value is the final
+     *               location for the uploaded file on the server.
+     */
+    public function getMovedFiles() : array
+    {
+        return $this->_movedFiles;
+    }
+
+    /**
+     * Accessor for field $_dirPermissions.
+     *
+     * @return int value for field $_dirPermissions.
+     */
+    public function getFiles() : array
+    {
+        if (is_null($this->_files)) {
+            $this->_files = array();
+            foreach ($this->_lorisForm->getElements() as $element) {
+                if ($element instanceof LorisFormFileElement) {
+                    $this->_files[$element->getName()] = $element->getValue();
+                }
+            }
+        }
+
+        return $this->_files;
+    }
+
+    /**
+     * Determines if the set of uploaded files is valid or not, updating the $_errors
+     * array with the errors found. This method will look at all the fields of the
+     * LorisForm associated to this handler and verify that the upload error code
+     * found in the super global $_FILES for that field is UPLOAD_ERR_OK
+     * (see documentation for $_FILES). If no file name was entered for a given field
+     * in the LorisForm, then the upload is considered invalid and the error for that
+     * field will be set to false. This method assumes that all the fields used for
+     * file uploads have a value (i.e. no "optional" upload is allowed). This
+     * behavior can be overridden bu subclasses should they need to.
+     *
+     * @return bool true if the uploaded file set is valid, false otherwise.
+     */
+    public function isValid() : bool
+    {
+        $valid = true;
+        foreach ($this->getFiles() as $field => $file) {
+            if ($file === false || $file['error'] != UPLOAD_ERR_OK) {
+                $this->addError(
+                    $field,
+                    $file === false ? false : $file['error']
+                );
+                $valid = false;
+            }
+        }
+
+        return $valid;
+    }
+
+    /**
+     * Main method for this handler. Validates the uploaded files (i.e. make sure
+     * that they are are present on the server) and moves them to their final
+     * locations.
+     *
+     * @return nothing.
+     */
+    public function processAll()
+    {
+        try {
```

Comment:
This will partially process the array and leave things in an inconsistent state if something in the middle throws an exception.

It's debatable whether the try should be moved inside the loop or not, but regardless the content of the loop should be moved to a `protected processFile($file) : bool ` (with some appropriate type for `$file`, or maybe throwing an exception instead of returning a bool) which is called for each file, so that a mechanism can be put in place for the programmer to determine how to deal with error cases.

Reply:
You do have access to the _errors and _movedFiles array so you could perform any "cleanup" in processingDone if anything goes seriously wrong (provided the exception is passed to processingDone, which is not the case right now). This would provide similar behaviour IMHO. I do see your point though. I am kind of on the fence regarding this...